ext/debug/ctf/ctf_middle.h

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2018 Oticon A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _CTF_MIDDLE_H
+#define _CTF_MIDDLE_H
+
+#include <stddef.h>
+#include <string.h>
+#include <ctf_bottom.h>
+
+
+/* Optionally enter into a critical region, decided by bottom layer */
+#define CTF_CRITICAL_REGION(x)	   \
+	{			   \
+		CTF_BOTTOM_LOCK;   \
+		x;		   \
+		CTF_BOTTOM_UNLOCK; \
+	}
+
+
+#ifdef CTF_BOTTOM_TIMESTAMPED_EXTERNALLY
+/* Emit CTF event using the bottom-level IO mechanics */
+#define CTF_EVENT(...)						    \
+	{							    \
+		CTF_CRITICAL_REGION(CTF_BOTTOM_FIELDS(__VA_ARGS__)) \
+	}
+#endif /* CTF_BOTTOM_TIMESTAMPED_EXTERNALLY */
+
+#ifdef CTF_BOTTOM_TIMESTAMPED_INTERNALLY
+/* Emit CTF event using the bottom-level IO mechanics. Prefix by sample time */
+#define CTF_EVENT(...)							    \
+	{								    \
+		const u32_t tstamp = k_cycle_get_32();			    \
+		CTF_CRITICAL_REGION(CTF_BOTTOM_FIELDS(tstamp, __VA_ARGS__)) \
+	}
+#endif /* CTF_BOTTOM_TIMESTAMPED_INTERNALLY */
+
+
+/* Anonymous compound literal with 1 member. Legal since C99.
+ * This permits us to take the address of literals, like so:
+ *  &CTF_LITERAL(int, 1234)
+ *
+ * This may be required if a ctf_bottom layer uses memcpy.
+ *
+ * NOTE string literals already support address-of and sizeof,
+ * so string literals should not be wrapped with CTF_LITERAL.
+ */
+#define CTF_LITERAL(type, value)  ((type){ (type)(value) })
+
+
+typedef enum {
+	CTF_EVENT_THREAD_SWITCHED_OUT   =  0x10,
+	CTF_EVENT_THREAD_SWITCHED_IN    =  0x11,
+	CTF_EVENT_THREAD_PRIORITY_SET   =  0x12,
+	CTF_EVENT_THREAD_CREATE         =  0x13,
+	CTF_EVENT_THREAD_ABORT          =  0x14,
+	CTF_EVENT_THREAD_SUSPEND        =  0x15,
+	CTF_EVENT_THREAD_RESUME         =  0x16,
+	CTF_EVENT_THREAD_READY          =  0x17,
+	CTF_EVENT_THREAD_PENDING        =  0x18,
+	CTF_EVENT_THREAD_INFO           =  0x19,
+	CTF_EVENT_ISR_ENTER             =  0x20,
+	CTF_EVENT_ISR_EXIT              =  0x21,
+	CTF_EVENT_ISR_EXIT_TO_SCHEDULER =  0x22,
+	CTF_EVENT_IDLE                  =  0x30,
+	CTF_EVENT_ID_START_CALL         =  0x41,
+	CTF_EVENT_ID_END_CALL           =  0x42
+} ctf_event_t;
+
+
+typedef struct {
+	char buf[20];
```

Comment:
Why 20?

Reply:
It seemed like a reasonable upper limit (4x32 bit words) for a string sent out to a potentially bandwidth limited backend. In general we should probably avoid sending strings as available bandwidth can be better used for other things.

Reply:
Latest push removes the magic number in favor of a proper define with comment explaining the reasoning.